client/src/main/scala/org/apache/celeborn/client/commit/CommitHandler.scala

```diff
@@ -180,6 +198,146 @@ abstract class CommitHandler(
+
+        // record failed partitions
+        shuffleCommittedInfo.failedPrimaryPartitionIds.putAll(
+          res.failedPrimaryIds.asScala.map((_, worker)).toMap.asJava)
+        shuffleCommittedInfo.failedReplicaPartitionIds.putAll(
+          res.failedReplicaIds.asScala.map((_, worker)).toMap.asJava)
+
+        shuffleCommittedInfo.committedMapIdBitmap.putAll(res.committedMapIdBitMap)
+
+        totalWritten.add(res.totalWritten)
+        fileCount.add(res.fileCount)
+        shuffleCommittedInfo.currentShuffleFileCount.add(res.fileCount)
+      }
+    }
+
+    val futures = new LinkedBlockingQueue[FutureWithStatus]()
+
+    val outFutures = params.filter(param =>
+      !CollectionUtils.isEmpty(param.primaryIds) ||
+        !CollectionUtils.isEmpty(param.replicaIds)) map { param =>
+      Future {
+        val future = commitFiles(
+          appUniqueId,
+          shuffleId,
+          param.worker,
+          param.primaryIds,
+          param.replicaIds)
+
+        futures.add(FutureWithStatus(future, param, 1))
+      }(ec)
+    }
+    val cbf =
+      implicitly[
+        CanBuildFrom[ArrayBuffer[Future[Boolean]], Boolean, ArrayBuffer[Boolean]]]
+    val futureSeq = Future.sequence(outFutures)(cbf, ec)
+    awaitResult(futureSeq, Duration.Inf)
+
+    val maxRetries = conf.clientRequestCommitFilesMaxRetries
+    var timeout = conf.rpcAskTimeout.duration.toMillis * maxRetries
+    val delta = 50
+    while (timeout >= 0 && !futures.isEmpty) {
+      val iter = futures.iterator()
+      while (iter.hasNext) {
+        val status = iter.next()
+        if (status.future.isCompleted) {
+          status.future.value.get match {
+            case scala.util.Success(res) =>
+              val worker = status.commitFilesParam.worker
+              res.status match {
+                case StatusCode.SUCCESS => // do nothing
+                case StatusCode.PARTIAL_SUCCESS | StatusCode.SHUFFLE_NOT_REGISTERED | StatusCode.REQUEST_FAILED | StatusCode.WORKER_EXCLUDED =>
+                  logInfo(s"Request commitFiles return ${res.status} for " +
+                    s"${Utils.makeShuffleKey(appUniqueId, shuffleId)}")
+                  if (res.status != StatusCode.WORKER_EXCLUDED) {
+                    commitFilesFailedWorkers.put(worker, (res.status, System.currentTimeMillis()))
+                  }
+                case _ =>
+                  logError(s"Should never reach here! commit files response status ${res.status}")
+              }
+
+              processResponse(res, worker)
+              iter.remove()
+            case scala.util.Failure(e) =>
+              val worker = status.commitFilesParam.worker
+              logError(
+                s"Ask worker($worker) CommitFiles for $shuffleId failed" +
+                  s" (attempt ${status.retriedTimes}/$maxRetries).",
+                e)
+              if (status.retriedTimes < maxRetries) {
```

Comment:
nit: as the actual default max retry times is`3`, change to `status.retriedTimes <= maxRetries` and set the default value of the conf `celeborn.client.requestCommitFiles.maxRetries` to 3?

Reply:
This is about whether we should count the first run in the "retry" times. If do, then the current logic is OK.

Reply:
then the `FutureWithStatus.retriedTimes` should start from `0` instead of `1`?

https://github.com/apache/incubator-celeborn/blob/3fb7edfa6970cae0a7e2c5d5108fc339884435a9/client/src/main/scala/org/apache/celeborn/client/commit/CommitHandler.scala#L257